### PR TITLE
Update ScyllaDB Monitoring to 4.8.1 and use versioned platform dashboards

### LINKED ISSUE #1505: Update ScyllaDBMonitoring
With ScyllaDB Monitoring 4.5.0 out, we should update operator dashboards and required grafana + prometheus versions.

https://forum.scylladb.com/t/release-scylladb-monitoring-stack-4-5-0/933

@amnonh can you please update the SaaS dashboard and bring in the rest?

fyi, the assets live in
https://github.com/scylladb/scylla-operator/tree/
59ff82e/assets/monitoring

### PR DESCRIPTION
**Pull request by tnozicka**
**Description of your changes:**
This PR replaces the current single platform dashboard we used to have for monitoring with versioned folders for each ScyllaDB version and manager versions supported by a particular monitoring release. Each of these folders is now backed by an independent ConfigMap to avoid the 1MiB size limit. (All of the dashboards are still gziped to fit in, at least by the folder.)

This brings in ScyllaDB Monitoring `v4.8.1`, Prometheus `v2.53.2`, Grafana `v11.1.7` and updates helper images as well.

ScyllaDB monitoring dashboards and rules can now be updated by checking out a new version in the submodule followed by `make update-monitoring`.

Given the changes and semi automatic updates, this also extends the testing coverage for monitoring to make these safer.

**Which issue is resolved by this Pull Request:**
Resolves #1505

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -75,6 +75,9 @@ CONTROLLER_GEN ?=$(GO) run ./vendor/sigs.k8s.io/controller-tools/cmd/controller-
 CRD_PATH ?= pkg/api/scylla/v1/scylla.scylladb.com_scyllaclusters.yaml
 CRD_FILES ?=$(shell find ./pkg/api/ -name '*.yaml')$(if $(filter $(.SHELLSTATUS),0),,$(error "can't find CRDs"))
 
+MONITORING_DASHBOARDS_DIR :=./submodules/github.com/scylladb/scylla-monitoring/grafana/build
+MONITORING_RULES_DIR :=./submodules/github.com/scylladb/scylla-monitoring/prometheus/prom_rules
```
Comment (rzetelskik) [inline Makefile:79]: How is the submodule meant to be synced? Manually?

File: Makefile [hunk 2]
```diff
@@ -552,6 +555,50 @@ verify-examples:
...
+	$(call embed-rules,$(1)/prometheus/v1/rules)
+
+endef
+
+update-monitoring: submodules
+	$(call embed-monitoring,./assets/monitoring)
+.PHONY: update-monitoring
+
+verify-monitoring: tmp_dir :=$(shell mktemp -d)
+verify-monitoring: submodules
```
Comment (rzetelskik) [inline Makefile:595]: there is no `submodules` target?

File: test/e2e/set/scylladbmonitoring/scylladbmonitoring.go [hunk 3]
```diff
@@ -161,25 +264,26 @@ var _ = g.Describe("ScyllaDBMonitoring", func() {
...
+			for _, ruleGroup := range rulesResult.Groups {
+				eo.Expect(ruleGroup.Name).To(o.Equal("scylla.rules"))
+				eo.Expect(ruleGroup.Rules).NotTo(o.BeEmpty())
+				for _, rule := range ruleGroup.Rules {
+					switch r := rule.(type) {
+					case promeheusappv1api.AlertingRule:
+						eo.Expect(r.Health).To(o.BeEquivalentTo(promeheusappv1api.RuleHealthGood))
+
+					case promeheusappv1api.RecordingRule:
+						eo.Expect(r.Health).To(o.BeEquivalentTo(promeheusappv1api.RuleHealthGood))
```
Comment (zimnx) [inline test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:280]:
```suggestion
					case promeheusappv1api.AlertingRule, promeheusappv1api.RecordingRule:
						eo.Expect(r.Health).To(o.BeEquivalentTo(promeheusappv1api.RuleHealthGood))
```

File: Makefile (hunk 1 above)
Comment (tnozicka) [reply Makefile:79]: manually, it's a git commit in the end

File: Makefile (hunk 2 above)
Comment (tnozicka) [reply Makefile:595]: looks like that got lost, thx
(it's just the init)

File: pkg/controller/scylladbmonitoring/sync_grafana.go [hunk 4]
```diff
@@ -147,33 +151,53 @@ func makeGrafanaConfigs(sm *scyllav1alpha1.ScyllaDBMonitoring) (*corev1.ConfigMa
 		enableAnonymousAccess = spec.Authentication.InsecureEnableAnonymousAccess
 	}
 
+	var defaultDashboard string
+	switch t := sm.Spec.GetType(); t {
+	case scyllav1alpha1.ScyllaDBMonitoringTypePlatform:
+		defaultDashboard = configassests.Project.Operator.GrafanaDefaultPlatformDashboard
+	case scyllav1alpha1.ScyllaDBMonitoringTypeSAAS:
+		defaultDashboard = "scylladb-latest/overview.json"
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/scylladbmonitoring/sync_grafana.go:159]: unless something else is accessing it I'd prefer not to

File: assets/monitoring/prometheus/v1/registry.go [hunk 5]
```diff
@@ -37,14 +39,22 @@ var (
...
+	LatencyPrometheusRuleTemplate       = ParseObjectTemplateOrDie[*monitoringv1.PrometheusRule]("latency-prometheus-rule", latencyPrometheusRuleTemplateString)
 
 	//go:embed "alerts.prometheusrule.yaml"
 	alertsPrometheusRuleTemplateString string
 	AlertsPrometheusRuleTemplate       = ParseObjectTemplateOrDie[*monitoringv1.PrometheusRule]("alerts-prometheus-rule", alertsPrometheusRuleTemplateString)
 
+	//go:embed "table.prometheusrule.yaml"
+	tablePrometheusRuleTemplateString string
+	TablePrometheusRuleTemplate       = ParseObjectTemplateOrDie[*monitoringv1.PrometheusRule]("table-prometheus-rule", tablePrometheusRuleTemplateString)
+
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/prometheus/v1/registry.go:57]: that's how they are in scylla-monitoring

File: assets/monitoring/grafana/v1alpha1/deployment.yaml [hunk 6]
```diff
@@ -31,9 +31,11 @@ spec:
...
         args:
         - |
-          find /var/run/configmaps -type d -mindepth 1 -maxdepth 1 | while read -r d; do
-            mkdir /var/run/decompressed-configmaps/"${d##*/}"
-            find "${d}" -maxdepth 1 -name '*.gz.base64' -exec cp -L -t /var/run/decompressed-configmaps/"${d##*/}" {} +
+          mkdir /var/run/decompressed-configmaps/grafana-scylladb-dashboards
+          find /var/run/configmaps -mindepth 2 -maxdepth 2 -type d | while read -r d; do
+            tp="/var/run/decompressed-configmaps/${d#"/var/run/configmaps/"}"
+            mkdir "${tp}"
+            find "${d}" -mindepth 1 -maxdepth 1 -name '*.gz.base64' -exec cp -L -t "${tp}" {} +
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/grafana/v1alpha1/deployment.yaml:38]: 6 lines bellow
https://github.com/scylladb/scylla-operator/pull/2144/files/ee7ebea5863792051953761c775a034092ac5021#diff-b52c7acafe342a51748a1e1c08e8b418476ac2dfc9831a46476d4c580747abf0R44

File: pkg/controller/scylladbmonitoring/sync_grafana.go [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
 
+	configassests "github.com/scylladb/scylla-operator/assets/config"
 	grafanav1alpha1assets "github.com/scylladb/scylla-operator/assets/monitoring/grafana/v1alpha1"
 	scyllav1alpha1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1alpha1"
 	"github.com/scylladb/scylla-operator/pkg/controllerhelpers"
 	"github.com/scylladb/scylla-operator/pkg/controllertools"
 	ocrypto "github.com/scylladb/scylla-operator/pkg/crypto"
 	"github.com/scylladb/scylla-operator/pkg/helpers"
-	"github.com/scylladb/scylla-operator/pkg/helpers/slices"
+	oslices "github.com/scylladb/scylla-operator/pkg/helpers/slices"
```
[the remark this answers is not in the document]
Comment (rzetelskik) [reply pkg/controller/scylladbmonitoring/sync_grafana.go:18]: my bad, I though it said `osslices` as in open source, oslices is fine

File: assets/monitoring/prometheus/v1/registry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rzetelskik) [reply assets/monitoring/prometheus/v1/registry.go:57]: ok

File: assets/monitoring/grafana/v1alpha1/grafanadashboards.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	grafanaDashboardsFileRegex = regexp.MustCompile(`^[^/]+\.json$`)
+)
+
+func gzipMapData(uncompressedMap map[string]string) (map[string]string, error) {
+	res := make(map[string]string, len(uncompressedMap))
+	for k, v := range uncompressedMap {
+		var buf bytes.Buffer
+		b64Writer := base64.NewEncoder(base64.StdEncoding, &buf)
+
+		zw, err := gzip.NewWriterLevel(b64Writer, gzip.BestCompression)
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/grafana/v1alpha1/grafanadashboards.go]: renamed

File: Makefile (hunk 2 above)
Comment (tnozicka) [reply Makefile:599]: added

**Comment by rzetelskik** [comment]
lgtm
/assign zimnx

File: assets/monitoring/grafana/v1alpha1/deployment.yaml [hunk 9]
```diff
@@ -43,8 +45,12 @@ spec:
...
         - name: decompressed-configmaps
           mountPath: /var/run/decompressed-configmaps
-        - name: grafana-scylladb-dashboards
-          mountPath: /var/run/configmaps/grafana-scylladb-dashboards
+        {{- range $_, $cm := .dashboardsCMs }}
+        - name: "{{ $cm.Name }}"
+          mountPath: "/var/run/configmaps/grafana-scylladb-dashboards/{{ index $cm.Annotations "internal.scylla-operator.scylladb.com/dashboard-name" }}"
+        {{- else }}
+        !!! No dashboards names found !!!
+        {{- end }}
```
[the remark this answers is not in the document]
Comment (zimnx) [reply assets/monitoring/grafana/v1alpha1/deployment.yaml:53]: I don't mind having both. One before rendering is missing.

File: test/e2e/set/scylladbmonitoring/scylladbmonitoring.go (hunk 3 above)
Comment (zimnx) [reply test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:280]: ah right, it's a type switch. I thought it's enum. nvm

**Review by zimnx — APPROVED** [review]
one open discussion, rest lgtm

File: assets/monitoring/grafana/v1alpha1/deployment.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/grafana/v1alpha1/deployment.yaml:53]: it's transitively here
https://github.com/scylladb/scylla-operator/blob/8c35cde94c7b7f62b591561f3eeb722364b6d667/assets/monitoring/grafana/v1alpha1/grafanadashboards_test.go#L26-L42
but I can add an extra one if that's the ask

[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/grafana/v1alpha1/deployment.yaml:53]: (the template can also be generally rendered at some other place, that's why I ended up with something that breaks the scheme as the most reliable while not the prettiest)

[the remark this answers is not in the document]
Comment (tnozicka) [reply assets/monitoring/grafana/v1alpha1/deployment.yaml:53]: i don't mind an extra check per say, added

**Review by zimnx — APPROVED** [review]
thanks for updates
/lgtm

**Pull request merged**